tools/kagami/src/main.rs

```diff
@@ -713,18 +717,44 @@ mod validator {
     impl<T: Write> RunArgs<T> for Args {
         fn run(self, writer: &mut BufWriter<T>) -> Outcome {
             writer
-                .write_all(&construct_validator()?)
+                .write_all(&construct_validator(None)?)
                 .wrap_err("Failed to write wasm validator into the buffer.")
         }
     }
 
-    pub fn construct_validator() -> color_eyre::Result<Vec<u8>> {
+    pub fn construct_validator(
+        swarm_build_dir: Option<&AbsolutePath>,
+    ) -> color_eyre::Result<Vec<u8>> {
         let build_dir = tempfile::tempdir()
             .wrap_err("Failed to create temp dir for runtime validator output")?;
+        let source_dir = tempfile::tempdir()
+            .wrap_err("Failed to create temp dir for default runtime validator sources")?;
+
+        // If `CARGO_MANIFEST_DIR` is set, it means the `kagami` binary is run from inside Iroha repo,
+        // otherwise we treat it as standalone, while also trying to avoid double repo cloning when
+        // called from `swarm`
+        let validator_path = std::env::var("CARGO_MANIFEST_DIR").map_or_else(
+            |_| {
+                if let Some(swarm_dir) = swarm_build_dir {
+                    Ok(swarm_dir.join("../../default_validator"))
+                } else {
+                    swarm::shallow_git_clone(
+                        swarm::GIT_ORIGIN,
+                        swarm::GIT_REVISION,
+                        &AbsolutePath::from(source_dir.path().join(swarm::DIR_CLONE)),
+                    )?;
+                    Ok(source_dir.path().to_owned())
+                }
+            },
+            |manifest_dir| {
+                Path::new(&manifest_dir)
+                    .join("../../default_validator")
+                    .canonicalize()
+                    .wrap_err("Failed to canonicalize path")
+            },
+        )?;
```

Comment:
You should pass `validator_path` as an argument and instead compute it from another function.